src/main/java/be/seeseemelk/mockbukkit/scheduler/BukkitSchedulerMock.java

```diff
@@ -38,9 +61,14 @@ public void shutdown()
 	{
 		waitAsyncTasksFinished();
 		pool.shutdown();
-
 		if (asyncException.get() != null)
 			throw new AsyncTaskException(asyncException.get());
+		chatExecutor.shutdownNow();
+	}
+
+	public void executeAsyncChatEvent(AsyncPlayerChatEvent event)
+	{
+		chatExecutor.submit(() -> MockBukkit.getMock().getPluginManager().callEvent(event));
```

Comment:
I would note that I added this event so that Chat is not executed using the standard scheduler pool and cause incrementing of the task count. 

However, it can be reverted in lieu of #187

Reply:
Oh my, I didn't even see this change.
I think running this on the executor's thread pool directly sounds like a good idea since this is just supposed to run on a different Thread directly, not necessarily going to have to go through all the scheduler logic first.
Especially when we consider that the scheduling of that event is only supposed to mock the multi-threaded nature of the chat protocol.

The only reason I even touched that part in #187 was to rectify the behaviour when events get triggered on the wrong Thread, since an exception should be thrown in that case.
We recently had this being the case, the unit test passed but when the project was put on a server, it spit out tons of exceptions due to events being scheduled on the wrong Thread.
So this fix could have saved us some trouble.

I think running this on the Executor directly is a good idea but I would generalize this to accept **any async event**.
Then we could have the `PluginManagerMock` run this in the background.
If you want, you could make a PR to my fork.
But I was gonna split that behemoth up into multiple pull requests anyway, so I could also just make a branch on this repo dedicated for this which you could then PR to.

Reply:
So the scheduler as it stands has 2 executors 
1. The Task Executor which handles all async task - the pool is exposed in this scheduler to allow monitoring of the actively running tasks.
2. The Chat executor - which is really fire and forget its purely internal for testing - and not representative of anything that runs on a live server.  Chat on live CAN be executed async but also sometimes isn't - in this case we artificially execute it both ways ie from the PlayerMock.class 
```java
@Override
	public void chat(String msg)
	{
		AsyncPlayerChatEvent eventAsync = new AsyncPlayerChatEvent(false, this, msg,
		        new HashSet<>(Bukkit.getOnlinePlayers()));
		PlayerChatEvent eventSync = new PlayerChatEvent(this, msg);
		MockBukkit.getMock().getScheduler().executeAsyncChatEvent(eventAsync);
		Bukkit.getPluginManager().callEvent(eventSync);
	}
```

Reply:
The only reason I put the Chat Executor in the scheduler mock is I could ensure both pools were shutdown at the same time.

Reply:
Ohh I see, forgive me I haven't had a chance to actually look at this code in detail yet.
Yeah that makes a lot of sense, personally I would probably have a dedicated `AsyncEventSchedulerMock` here though which we could just have as yet another final field in the Scheduler. But that may just be my personal preference.

Also note that any Event can be asynchronous, not just the `AsyncPlayerChatEvent`, even though it is the most prominent implementation (perhaps even the only one in vanilla Bukkit). So the naming should definitely reflect that as "Chat" could be misleading then.

I should probably properly look into this pull request next week (busy week right now).
But currently I am thinking it would be best to seperate out the async event handling from both of our pull requests and work on one dedicated PR for that together, since we are both addressing different problems that are related to this and a combined solution would probably be for the better.

Reply:
I'm not sure that it is a good idea to execute the event twice. That means any event handlers would also get executed twice, which I don't think is something we want to happen.

Reply:
That was the existing behaviour